File: src/DMIslandClient/Connection/Dto.fs
namespace DMIslandClient.Connection

open System
open System.Collections.Generic
open System.Text.Json.Serialization

module Dto =
    type EventType =
        | EntityDeath = 1
        | TearPop = 2
        | EnemyProjectilePop = 3

    type EntityType =
        | Player = 1
        | ModusPonens = 2
        | Lambda = 3
        | Monad = 4
        | Tear = 5
        | EnemyProjectile = 6
        | Wall = 7
        | HeartItem = 8
        | HalfHeartItem = 9
        | AmethystItem = 10
        | CppItem = 11
        | HaskellItem = 12
        | Python3Item = 13
        | JavaItem = 14
        | OCamlItem = 15
        | ZigItem = 16
        | RustItem = 17
        | AnsiCItem = 18
        | FSharpItem = 19
        | RocItem = 20
        | OneFItem = 21
        | JavaScriptItem = 22
        | TypeScriptItem = 23
        | GoItem = 24
        | KotlinItem = 25
        | AsmItem = 26
        | Scala3Item = 27
        | Nerd = 28
        | NuclearNerd = 29
        | Skolem = 30
        | Mole = 31
        | AttackIndicator = 32
        | ThetaAttack = 33

    type ItemType =
        | Heart = 1
        | HalfHeart = 2
        | Amethyst = 3
        
        | Cpp = 4
        | Haskell = 5
        | Python3 = 6
        | Java = 7
        | OCaml = 8
        | Zig = 9
        | Rust = 10
        | AnsiC = 11
        | FSharp = 12
        | Roc = 13
        | OneF = 14
        | JavaScript = 15
        | TypeScript = 16
        | Go = 17
        | Kotlin = 18
        | Asm = 19
        | Scala3 = 20
    
    [<CLIMutable>]
    type PlayerActionRequest = {
        [<JsonPropertyName("action")>]
        Action: string
        
        [<JsonPropertyName("direction")>]
        Direction: string option
    }

    [<CLIMutable>]
    type PositionDto = {
        [<JsonPropertyName("x")>]
        X: int
        
        [<JsonPropertyName("y")>]
        Y: int
    }
    
    [<CLIMutable>]
    type RoomDto = {
        [<JsonPropertyName("id")>]
        Id: Guid
        
        [<JsonPropertyName("biome")>]
        Biome: String
    }

    [<CLIMutable>]
    type ObjectViewDto = {
        [<JsonPropertyName("id")>]
        Id: Guid
        
        [<JsonPropertyName("type")>]
        [<JsonConverter(typeof<JsonStringEnumConverter>)>]
        Type: EntityType
        
        [<JsonPropertyName("name")>]
        Name: string
        
        [<JsonPropertyName("hp")>]
        Hp: int
        
        [<JsonPropertyName("maxHp")>]
        MaxHp: int
        
        [<JsonPropertyName("position")>]
        Position: PositionDto
        
        [<JsonPropertyName("previous_position")>]
        PreviousPosition: PositionDto
    }
    
    [<CLIMutable>]
    type EventDto = {
        [<JsonPropertyName("type")>]
        [<JsonConverter(typeof<JsonStringEnumConverter>)>]
        Type: EventType
        
        [<JsonPropertyName("position")>]
        Position: PositionDto
        
        [<JsonPropertyName("payload")>]
        Payload: string
    }

    [<CLIMutable>]
    type RoomCellDto = {
        [<JsonPropertyName("x")>]
        X: int

        [<JsonPropertyName("y")>]
        Y: int

        [<JsonPropertyName("visited")>]
        Visited: bool

        [<JsonPropertyName("cleared")>]
        Cleared: bool

        [<JsonPropertyName("current")>]
        Current: bool
    }

    [<CLIMutable>]
    type GameStateResponse = {
        [<JsonPropertyName("turn")>]
        Turn: int

        [<JsonPropertyName("player")>]
        Player: ObjectViewDto

        [<JsonPropertyName("objects")>]
        Objects: List<ObjectViewDto>
        
        [<JsonPropertyName("entities")>]
        Entities: List<ObjectViewDto>

        [<JsonPropertyName("viewWidth")>]
        ViewWidth: int

        [<JsonPropertyName("viewHeight")>]
        ViewHeight: int

        [<JsonPropertyName("events")>]
        Events: List<EventDto>

        [<JsonPropertyName("floor")>]
        Floor: int
        
        [<JsonPropertyName("items")>]
        Items: List<ItemType>

        [<JsonPropertyName("completed")>]
        Completed: bool

        [<JsonPropertyName("room")>]
        Room: RoomDto

        [<JsonPropertyName("rooms")>]
        Rooms: List<RoomCellDto>
    }

File: services/DMIslandServer/Api/Dto.cs
using System;
using System.Collections.Generic;
using System.Text.Json.Serialization;
using RoguelikeServerMVP.Game;

namespace RoguelikeServerMVP.Api;

public class PlayerActionRequest
{
    [JsonPropertyName("action")]
    public string Action { get; set; } = string.Empty;

    [JsonPropertyName("direction")]
    public string? Direction { get; set; }
}

public class PositionDto
{
    [JsonPropertyName("x")]
    public int X { get; set; }

    [JsonPropertyName("y")]
    public int Y { get; set; }

    public PositionDto(int x, int y)
    {
        X = x;
        Y = y;
    }

    public PositionDto(Position position)
    {
        X = position.X;
        Y = position.Y;
    }
}

public class RoomDto
{
    [JsonPropertyName("id")]
    public Guid Id { get; set; }

    [JsonPropertyName("biome")]
    public string Biome { get; set; } = "beach";

    [JsonPropertyName("width")]
    public int Width { get; set; }

    [JsonPropertyName("height")]
    public int Height { get; set; }
}

public class ObjectViewDto
{
    [JsonPropertyName("id")]
    public Guid Id { get; set; }

    [JsonPropertyName("type")]
    [JsonConverter(typeof(JsonStringEnumConverter))]
    public EntityType Type { get; set; }

    [JsonPropertyName("hp")]
    public int Hp { get; set; }

    [JsonPropertyName("maxHp")]
    public int MaxHp { get; set; }

    [JsonPropertyName("position")]
    public PositionDto Position { get; set; }

    [JsonPropertyName("previous_position")]
    public PositionDto PreviousPosition { get; set; }
}

public class GameStateResponse
{
    [JsonPropertyName("turn")]
    public int Turn { get; set; }

    [JsonPropertyName("player")]
    public ObjectViewDto Player { get; set; } = new();

    [JsonPropertyName("entities")]
    public List<ObjectViewDto> Entities { get; set; } = [];
    
    [JsonPropertyName("objects")]
    public List<ObjectViewDto> Objects { get; set; } = [];
    
    [JsonPropertyName("events")]
    public List<EventDto> Events { get; set; } = [];

    [JsonPropertyName("viewWidth")]
    public int ViewWidth { get; set; }

    [JsonPropertyName("viewHeight")]
    public int ViewHeight { get; set; }

    [JsonPropertyName("floor")]
    public int Floor { get; set; }

    [JsonPropertyName("completed")]
    public bool Completed { get; set; }
    
    [JsonPropertyName("items")]
    public List<ItemType>? Items { get; set; }

    [JsonPropertyName("room")]
    public RoomDto Room { get; set; }

    [JsonPropertyName("rooms")]
    public List<RoomCellDto> Rooms { get; set; } = [];
}

public class RoomCellDto
{
    [JsonPropertyName("x")]
    public int X { get; set; }

    [JsonPropertyName("y")]
    public int Y { get; set; }

    [JsonPropertyName("visited")]
    public bool Visited { get; set; }

    [JsonPropertyName("cleared")]
    public bool Cleared { get; set; }

    [JsonPropertyName("current")]
    public bool Current { get; set; }

    /// <summary>True when this room holds the (active) floor exit — used to mark it on the minimap.</summary>
    [JsonPropertyName("isExit")]
    public bool IsExit { get; set; }
}

public class EventDto
{
    [JsonPropertyName("type")]
    [JsonConverter(typeof(JsonStringEnumConverter))]
    public EventType Type { get; set; }

    [JsonPropertyName("position")]
    public Position Position { get; set; } = Position.Zero;
    
    [JsonPropertyName("payload")]
    public string Payload { get; set; } = string.Empty;
}


public enum EventType
{
    PlayerDied,
    EntityDeath,
    EnemyProjectilePop,
    TearPop
}

public enum EntityType
{
    Player, 
    
    ModusPonens,
    Lambda,
    Monad,
    Nerd,
    NuclearNerd,
    Skolem,
    Mole,
    Tear,
    
    AttackIndicator,
    ThetaAttack,
    Lightning,
    EnemyProjectile,
    Wall,
    Exit,
    ExitClosed,

    HeartItem,
    HalfHeartItem,
    AmethystItem,

    CppItem,
    HaskellItem,
    Python3Item,
    JavaItem,
    OCamlItem,
    ZigItem,
    RustItem,
    AnsiCItem,
    FSharpItem,
    RocItem,
    OneFItem,
    JavaScriptItem,
    TypeScriptItem,
    GoItem,
    KotlinItem,
    AsmItem,
    Scala3Item
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum ItemType
{
    Heart,
    HalfHeart,
    Amethyst,
    
    Cpp, 
    Haskell,
    Python3,
    Java,
    OCaml,
    Zig,
    Rust,
    AnsiC,
    FSharp,
    Roc,
    OneF,
    JavaScript,
    TypeScript,
    Go,
    Kotlin,
    Asm,
    Scala3
}